ci/cron/src/Docs.hs

```diff
@@ -76,10 +97,18 @@ build_and_push :: DocOptions -> FilePath -> [Version] -> IO ()
 build_and_push opts@DocOptions{build} temp versions = do
     restore_sha $ do
         Data.Foldable.for_ versions (\version -> do
-            putStrLn $ "Building " <> show version <> "..."
-            build temp version
-            putStrLn $ "Pushing " <> show version <> " to S3 (as subfolder)..."
-            push version
+            putStrLn $ "Check if version  " <> show version <> " exists ..."
+            -- We use a check for the versions.json file to check if the docs exists or not.
+            -- This is technically slightly racy if the upload happens concurrently and we end up with a partial upload
+            -- but the window seems small enough to make this acceptable.
+            r <- tryIO $ IO.withTempFile $ \file -> proc_ ["aws", "s3", "cp", s3Path opts (show version </> "versions.json"), file]
```

Comment:
I'm a bit confused as to why we need to add this check. What failure scenario is this preventing? As I understand it, for this check to change the current behaviour it would require that a release exists on GitHub while simultaneously not being present in either snapshots.json or versions.json (because if the version is in either of those we never call `build_and_push` for that version) but having its folder already present on S3.

In other words, that would require the assembly repo to create the folder on S3, and then either completely forget about snapshots.json or for some reason choose to update snapshots.json _after_ updating the GitHub version listing (and this to run in-between).

It seems to me the simpler fix is to program the assembly repo to only create the GitHub version entry _after_ it has both pushed the version to its subfolder and updated snapshots.json.

Perhaps we should instead worry about writes to snapshots.json not being transactional? As far as I can tell, S3 does not have a "compare and swap" feature. That wasn't a problem with only one script (this one) pushing files, but if both this and the assembly repo try to write snapshots.json we may have race conditions on that.

Reply:
I’d only like the assembly repo to push the docs to the versioned directory in S3. It will never update the snapshots.json, that’s done by this cron job. I’m using the versions.json file as a proxy for whether the whole directory exists or not. Maybe there is a better way to do this but this seemed good enough.

This check is to differentiate between split releases where the docs are already uploaded and non-split releases where we have to build & upload in the cron job. We could do that based on a version guard but it seemed like it is a bit nicer to allow us to run them side by side.

does that make sense? Happy to do something else if you have a better idea.

Reply:
Right. Your way is better, let's do that.

To clarify: we keep this PR as-is and the assembly repo uploads the `$version` subfolder without touching the snapshots.json file.

Reply:
Could this one line be a function call, though? Something like `does_version_exist` would help readability (even as a local `where` definition).

Reply:
apologies, didn’t mean to ignore your comment just merged too quickly https://github.com/digital-asset/daml/pull/12459